src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -864,7 +828,12 @@ void StorageReplicatedMergeTree::drop()
     /// or metadata of staled replica were removed manually,
     /// in this case, has_metadata_in_zookeeper = false, and we also permit to drop the table.
 
-    bool maybe_has_metadata_in_zookeeper = !has_metadata_in_zookeeper.has_value() || *has_metadata_in_zookeeper;
+    bool maybe_has_metadata_in_zookeeper = false;
+    {
+        std::lock_guard lock{initialization_mutex};
+        maybe_has_metadata_in_zookeeper = !initialization_done || !has_metadata_in_zookeeper.has_value() || *has_metadata_in_zookeeper;
```

Comment:
Do we really need to check `initialization_done` here? Seems like `!has_metadata_in_zookeeper.has_value()` is enough.

Reply:
But `has_metadata_in_zookeeper` isn't an atomic and can change in parallel in AttachThread.
Am I missing something?

Reply:
`shutdown()` must be called before `drop()` and `shutdown()` stops AttachThread, so it cannot change in parallel

Reply:
okay
then the `shutdown` call in `drop` gives a different impression.
Should that be an assert then? (that shutdown was called)

Reply:
Hm, this `shutdown()` call is from 2014 and it's not clear why it was added (https://github.com/ClickHouse/ClickHouse/commit/b4d85c1676cda6ee9753ec2ccc69b5784907008b). AFAIK, now we always call `shutdown()` before dropping (or detaching) a table

> Should that be an assert then? (that shutdown was called)

Yep

Reply:
Sorry, I forgot about this case. In this case we don't need to call `shutdown()` because `startup()` was not called (so either AttachThread was not started at all or it was already stopped)